wrappers/services.go

```diff
@@ -75,30 +76,109 @@ func generateSnapServiceFile(app *snap.AppInfo, opts *AddSnapServicesOptions) ([
 	return genServiceFile(app, opts)
 }
 
+// max returns the maximum of two integers. Why is this
+// not provided by golang?
+func max(a, b int) int {
+	if a < b {
+		return b
+	}
+	return a
+}
+
+func min(a, b int) int {
+	if a > b {
+		return b
+	}
+	return a
+}
+
+func formatCpuGroupSlice(grp *quota.Group) string {
+	if grp.CpuLimit == nil {
+		return ""
+	}
+
+	// calculateSystemdCPULimit calculates the number of cpus and the allowed percentage
+	// to the systemd specific format. The CPUQuota setting is only available since systemd 213
+	calculateSystemdCPULimit := func() string {
+		if grp.CpuLimit.Count == 0 && grp.CpuLimit.Percentage == 0 {
+			return ""
+		}
+
+		cpuQuotaFormat := "CPUQuota=%d%%\n"
+		cpuQuotaSnap := max(grp.CpuLimit.Count, 1) * grp.CpuLimit.Percentage
+		cpuQuotaMax := runtime.NumCPU() * 100
+		return fmt.Sprintf(cpuQuotaFormat, min(cpuQuotaSnap, cpuQuotaMax))
+	}
+
+	// getAllowedCpusValue converts allowed CPUs array to a comma-delimited
+	// string that systemd expects it to be in. If the array is empty then
+	// an empty string is returned
+	getAllowedCpusValue := func() string {
+		if len(grp.CpuLimit.AllowedCpus) == 0 {
+			return ""
+		}
+
+		allowedCpusFormat := "AllowedCPUs=%s\n"
+		allowedCpusValue := strings.Trim(strings.Join(strings.Fields(fmt.Sprint(grp.CpuLimit.AllowedCpus)), ","), "[]")
+		return fmt.Sprintf(allowedCpusFormat, allowedCpusValue)
+	}
+
+	// AllowedCpus is only available for cgroupsv2
+	template := `# Always enable cpu accounting, so the following cpu quota options have an effect
+CPUAccounting=true
+`
+	return fmt.Sprint(template, calculateSystemdCPULimit(), getAllowedCpusValue(), "\n")
+}
+
+func formatMemoryGroupSlice(grp *quota.Group) string {
+	if grp.MemoryLimit == 0 {
+		return ""
+	}
+
+	template := `# Always enable memory accounting otherwise the MemoryMax setting does nothing.
+MemoryAccounting=true
+MemoryMax=%[1]d
+# for compatibility with older versions of systemd
+MemoryLimit=%[1]d
```

Comment:
Are older systemd version fine with MemoryMax, or do you need to check the systemd version in order to figure out if you can set it?

Reply:
'MemoryMax' will provide a warning in the log in the forms of "Unknown lvalue 'MemoryMax' in section 'Service' " by systemd on older versions, but AFAIK it will not cause any issues other than that warning.

Reply:
yes this warning is okay, we have done this for other things too